File: core/src/main/java/net/cjsah/bot/plugin/MainPlugin.java
package net.cjsah.bot.plugin;

import net.cjsah.bot.Main;
import net.cjsah.bot.SignalType;
import net.cjsah.bot.command.Commands;
import net.cjsah.bot.command.simple.SimpleCommand;
import net.cjsah.bot.command.source.CommandSource;
import net.cjsah.bot.data.GroupUserData;
import net.cjsah.bot.data.UserData;
import net.cjsah.bot.event.EventManager;
import net.cjsah.bot.event.events.FriendMessageEvent;
import net.cjsah.bot.event.events.GroupMessageEvent;
import net.cjsah.bot.event.events.HeartbeatEvent;
import net.cjsah.bot.event.events.LifecycleEvent;
import net.cjsah.bot.permission.PermissionRole;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Collections;

public final class MainPlugin extends Plugin {
    private static final Logger log = LoggerFactory.getLogger("Console");
    public static final MainPlugin INSTANCE = new MainPlugin();
    public static final PluginInfo PLUGIN_INFO = new PluginInfo("core", "Core", "Core Plugin", "1.0", Collections.singletonMap("authors", Collections.singletonList("Cjsah")));

    @Override
    public void onLoad() {
        Commands.registerContext().register(MainPlugin.class);

        String pluginId = PLUGIN_INFO.getId();

//        @Deprecated
//        EventManager.subscribe(pluginId, MessageEvent.class, event ->
//                log.info("[{}] [{}] [{}({})] => {}", event.getRoomName(), event.getChannelName(), event.getUserName(), event.getUserId(), event.getMsg())
//        );

//        EventManager.subscribe(pluginId, MessageEmojiPinEvent.class, event -> {
//            log.info("{} {} {} {}", event.getUserId(), event.getEmoji(), event.getChannelId(), event.getMsgId());
//        });

        EventManager.subscribe(pluginId, LifecycleEvent.class, event -> {
            if (event.getStatus() == LifecycleEvent.Status.CONNECT) {
                Main.lifecycle(false, 0);
            }
        });

        EventManager.subscribe(pluginId, HeartbeatEvent.class, event -> Main.lifecycle(true, event.getInterval()));

        EventManager.subscribe(pluginId, FriendMessageEvent.class, event -> {
            UserData sender = event.getSender();
            log.info("[{}] [{}({})] => {}", event.getMode().getType(), sender.getNickname(), sender.getUserId(), event.getMessage());
        });

        EventManager.subscribe(pluginId, GroupMessageEvent.class, event -> {
            GroupUserData sender = event.getSender();
            log.info("[群聊] [{}({})] [{}({})] => {}", event.getGroupName(), event.getGroupId(), sender.getCard(), sender.getUserId(), event.getMessage());
        });

//        EventManager.subscribe(pluginId, CommandEvent.class, event -> {
////            log.info("[{}({})] [{}({})] ==> 触发命令: /{}",
////                    event.getRoomInfo().getName(),
////                    event.getRoomInfo().getId(),
////                    event.getSenderInfo().getNickname(),
////                    event.getSenderInfo().getId(),
////                    event.getCommandInfo().getCommand()
////            );
//            CommandSource source = new CommandSource(event);
//            CommandManager.execute(event.getCommandInfo(), source);
//        });
    }

    @SimpleCommand(value = "/botstop", permission = PermissionRole.ADMIN)
    public static void botStop(CommandSource<?> source) {
        source.sendFeedback("bot正在关闭...");
        Main.sendSignal(SignalType.STOP);
    }

    @SimpleCommand(value = "/test", permission = PermissionRole.OWNER)
    public static void test(CommandSource<?> source) {
        Main.sendSignal(SignalType.RESTART);
    }

}

File: core/src/main/kotlin/net/cjsah/bot/plugin/MainPlugin.kt
package net.cjsah.bot.plugin

import net.cjsah.bot.Signal
import net.cjsah.bot.api.Api
import net.cjsah.bot.command.CommandManager
import net.cjsah.bot.command.source.ConsoleCommandSource
import net.cjsah.bot.command.source.GroupCommandSource
import net.cjsah.bot.command.source.UserCommandSource
import net.cjsah.bot.data.GroupSourceData
import net.cjsah.bot.event.EventManager
import net.cjsah.bot.event.events.AppHeartBeatEvent
import net.cjsah.bot.event.events.GroupUserJoinEvent
import net.cjsah.bot.event.events.GroupUserLeaveEvent
import net.cjsah.bot.event.events.MessageEvent
import net.cjsah.bot.heart
import net.cjsah.bot.log
import java.util.*

class MainPlugin : Plugin() {

    override fun onLoad() {
        EventManager.subscribe(INSTANCE, AppHeartBeatEvent::class.java) {
            Signal.fromStatus(it.status)
            heart?.heart(it.interval)
        }

        EventManager.subscribe(INSTANCE, MessageEvent.GroupMessageEvent::class.java) {
            log.info("[群] [${it.groupId}] [${it.userId}(${it.sender.card})] => ${it.message}")
            if (it.rawMessage.startsWith("/")) {
                val group = Api.getGroupInfo(it.groupId, false)
                val user = Api.getGroupMemberInfo(it.groupId, it.userId, false)
                val source = GroupCommandSource(GroupSourceData(group, user))
                CommandManager.execute(it.rawMessage.substring(1), source)
            }
        }

        EventManager.subscribe(INSTANCE, MessageEvent.FriendMessageEvent::class.java) {
            log.info("[好友] [${it.userId}(${it.sender.nickname})] => ${it.message}")
            if (it.rawMessage.startsWith("/")) {
                val source = UserCommandSource(it.sender)
                CommandManager.execute(it.rawMessage.substring(1), source)
            }
        }

        EventManager.subscribe(INSTANCE, GroupUserJoinEvent::class.java) {
            log.info("[群] ${it.userId} 加入群 [${it.groupId}]")
        }

        EventManager.subscribe(INSTANCE, GroupUserLeaveEvent::class.java) {
            log.info("[群] ${it.userId} 离开了群 [${it.groupId}]")
        }

        CommandManager.register { dispatcher ->
            dispatcher.register(CommandManager.literal("console").then(CommandManager.literal("stop").executes("关闭Bot") {
                Signal.stop()
            }))
            dispatcher.register(CommandManager.literal("help").executes("帮助") { context ->
                val helps = dispatcher.getHelp(context.source)
                if (context.source is ConsoleCommandSource) {
                    helps.entries.forEach { context.source.sendFeedback("${it.key}\t${it.value}") }
                } else {
                    val collect = helps.entries.joinToString("\n") { "${it.key}\t${it.value}" }
                    context.source.sendFeedback(collect)
                }
            })

//            dispatcher.register(CommandManager.literal("test").then(CommandManager.literal("xml").executes("测试xml消息") { context ->
//                val text = FileUtil.readUtf8String(File("test.xml"))
//                val msg = MessageChain.of(XMLMessageNode(text))
//                context.source.sendFeedback(msg)
//            }).then(CommandManager.literal("json").executes("测试json消息") { context ->
//                val text = FileUtil.readUtf8String(File("test.json"))
//                val msg = MessageChain.of(JsonMessageNode(text))
//                println(msg.toJson())
//                context.source.sendFeedback(msg)
//            }))

        }
//
//        EventManager.subscribe((INSTANCE), MessageEvent.GroupMessageEvent::class.java) {
//            if (it.groupId == 799652476L && it.rawMessage == "/api") {
//                Api.sendGroupMsg(it.groupId, MessageChain.raw("此消息为机器人发送"))
//            }
//
//        }
    }

    companion object {
        @JvmField
        val INSTANCE = MainPlugin()
        @JvmField
        val PLUGIN_INFO = PluginInfo(
            "main",
            "Main",
            "Main",
            "1.0",
            Collections.singletonMap<String, Any>("authors", listOf("Cjsah"))
        )
    }

}